typing/env.ml

```diff
@@ -1812,77 +1837,87 @@ and store_type ~check id info env =
+        Warnings.is_active (Warnings.Unused_constructor ("", Unused))
+      then begin
+        let ty_name = Ident.name id in
+        let priv = info.type_private in
+          List.iter
+            begin fun (_, cstr) ->
+              let name = cstr.cstr_name in
+              let loc = cstr.cstr_loc in
+              let k = cstr.cstr_uid in
+              if not (Types.Uid.Tbl.mem !used_constructors k) then
+                let used = constructor_usages () in
+                Types.Uid.Tbl.add !used_constructors k
+                  (add_constructor_usage used);
+                if not (ty_name = "" || ty_name.[0] = '_')
+                then !add_delayed_check_forward
+                    (fun () ->
+                      Option.iter
+                        (fun complaint ->
+                           if not (is_in_signature env) then
+                             Location.prerr_warning loc
+                               (Warnings.Unused_constructor(name, complaint)))
+                        (constructor_usage_complaint ~rebind:false priv used))
+            end
+            constructors
+      end;
+      Type_variant (List.map snd constructors),
+      { env with
+        constrs =
+          List.fold_right
+            (fun (id, descr) constrs ->
+              let cda = { cda_description = descr; cda_address = None } in
+              TycompTbl.add id cda constrs)
+            constructors env.constrs;
+      }
+  | Type_record (_, repr) ->
+      let labels = Datarepr.labels_of_type path info in
+      if check && not loc.Location.loc_ghost &&
+        Warnings.is_active (Warnings.Unused_field ("", Unused))
+      then begin
+        let ty_name = Ident.name id in
+        let priv = info.type_private in
+        List.iter
+          begin fun (_, lbl) ->
+            let name = lbl.lbl_name in
+            let loc = lbl.lbl_loc in
+            let mut = lbl.lbl_mut in
+            let k = lbl.lbl_uid in
+            if not (Types.Uid.Tbl.mem !used_labels k) then
+              let used = label_usages () in
+              Types.Uid.Tbl.add !used_labels k
+                (add_label_usage used);
+              if not (ty_name = "" || ty_name.[0] = '_' || name.[0] = '_')
+              then !add_delayed_check_forward
+                  (fun () ->
+                    Option.iter
+                      (fun complaint ->
+                         if not (is_in_signature env) then
+                           Location.prerr_warning
+                             loc (Warnings.Unused_field(name, complaint)))
+                      (label_usage_complaint priv mut used))
+          end
+          labels
+      end;
+      Type_record (List.map snd labels, repr),
+      { env with
+        labels =
+          List.fold_right
+            (fun (id, descr) labels -> TycompTbl.add id descr labels)
+            labels env.labels;
+      }
+  | Type_abstract -> Type_abstract, env
+  | Type_open -> Type_open, env
```

Comment:
This is another `bimap` call followed by a pattern match to update `env` and a final `bimap` to remove the extra infos. (The probably code is probably simpler to follow).